core/src/main/java/org/infinispan/jmx/AbstractJmxRegistration.java

```diff
@@ -18,65 +22,231 @@
 import org.infinispan.util.logging.LogFactory;
 
 /**
- * Parent class for top level JMX component registration.
+ * Parent class for JMX component registration.
  *
  * @author Galder Zamarreño
  * @since 4.0
  */
 @Scope(Scopes.NONE)
-public abstract class AbstractJmxRegistration {
+abstract class AbstractJmxRegistration {
+
    private static final Log log = LogFactory.getLog(AbstractJmxRegistration.class);
 
-   protected ComponentsJmxRegistration registrar;
+   static final String NAME_KEY = "name";
 
-   @Inject GlobalConfiguration globalConfig;
+   private static final String COMPONENT_KEY = "component";
 
-   @Inject BasicComponentRegistry basicComponentRegistry;
+   @Inject
+   GlobalConfiguration globalConfig;
+
+   @Inject
+   BasicComponentRegistry basicComponentRegistry;
 
-   String jmxDomain;
    MBeanServer mBeanServer;
 
-   protected abstract ComponentsJmxRegistration buildRegistrar();
+   String jmxDomain;
+
+   private String groupName;
+
+   /**
+    * Looks up the MBean server and initializes domain and group. Must be called by subclasses on start if JMX is
+    * enabled (see {@link #isJMXEnabled()}.
+    */
+   final synchronized void initJMX() {
+      // prevent double lookup of MBeanServer on restart
+      if (mBeanServer == null) {
+         try {
+            GlobalJmxStatisticsConfiguration globalJmxConfig = globalConfig.globalJmxStatistics();
+            MBeanServerLookup lookup = globalJmxConfig.mbeanServerLookup();
+            if (lookup != null) {
+               mBeanServer = lookup.getMBeanServer(globalJmxConfig.properties());
+            }
+         } catch (Exception e) {
+            log.debug("Ignoring exception in MBean server lookup", e);
+         }
+      }
+
+      if (mBeanServer != null) {
+         // prevent double init of domain and group on restart
+         if (groupName == null) {
+            groupName = initGroup();
+         }
+         if (jmxDomain == null) {
+            jmxDomain = initDomain();
+         }
+      }
+   }
+
+   public abstract boolean isJMXEnabled();
+
+   /**
+    * Subclasses must implement this hook to initialize {@link #jmxDomain}.
+    */
+   protected abstract String initDomain();
+
+   /**
+    * Subclasses must implement this hook to initialize {@link #groupName}.
+    */
+   protected abstract String initGroup();
+
+   /**
+    * Gets the domain name. This should not be called unless JMX is enabled.
+    */
+   public final String getDomain() {
+      if (mBeanServer == null) {
```

Comment:
I think either `mBeanServer` has to be `volatile` or all of the methods with `mBeanServer == null` checks need to be synchronized to ensure the var initialization in `initJMX` is propogated across multiple threads.

Reply:
I'm leaning towards synchronized init. Will do that ASAP.

Reply:
I still think this needs addressing unless we can guarantee that calls to `getGroupName` etc are on the same thread as calls to `init`.  

We need the `mBeanServer` to be volatile, as well as the synchronized block in init`, so that the `mBeanServer == null` non-synchornized checks in `getGroupName` etc see the latest value.

Reply:
Made those volatile also. Should be good now